olp-cpp-sdk-authentication/src/AuthenticationClientImpl.cpp

```diff
@@ -136,112 +148,149 @@ SignInResult AuthenticationClientImpl::ParseAuthResponse(
+        auto it = cache.Find(credentials.GetKey());
+        return it != cache.end() ? boost::make_optional(it->value())
+                                 : boost::none;
+      });
+}
+
+template <>
+void AuthenticationClientImpl::StoreInCache(
+    const AuthenticationCredentials& credentials, SignInResult response) {
+  // Cache the response
+  client_token_cache_->locked(
+      [&](utils::LruCache<std::string, SignInResult>& cache) {
+        return cache.InsertOrAssign(credentials.GetKey(), response);
+      });
+}
 
-  return *cached_response;
+template <>
+void AuthenticationClientImpl::StoreInCache(
+    const AuthenticationCredentials& credentials, SignInUserResult response) {
+  // Cache the response
+  user_token_cache_->locked(
+      [&](utils::LruCache<std::string, SignInUserResult>& cache) {
+        return cache.InsertOrAssign(credentials.GetKey(), response);
+      });
 }
 
 client::CancellationToken AuthenticationClientImpl::SignInClient(
     AuthenticationCredentials credentials, SignInProperties properties,
     SignInClientCallback callback) {
-  auto sign_in_task =
-      [=](client::CancellationContext context) -> SignInClientResponse {
-    {
-      if (!settings_.network_request_handler) {
-        return {{static_cast<int>(http::ErrorCode::IO_ERROR),
-                 "Cannot sign in while offline"}};
-      }
+  auto task = [=](client::CancellationContext context) -> SignInClientResponse {
+    if (!settings_.network_request_handler) {
+      return client::ApiError::NetworkConnection(
+          "Cannot sign in while offline");
+    }
 
-      if (context.IsCancelled()) {
-        return {
-            {static_cast<int>(http::ErrorCode::CANCELLED_ERROR), "Cancelled"}};
+    if (context.IsCancelled()) {
+      return client::ApiError::Cancelled();
+    }
+
+    client::OlpClient client = CreateOlpClient(settings_, boost::none, 0);
+
+    std::time_t timestamp =
+        std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
+
+    if (!settings_.use_system_time) {
+      auto server_time = GetTimeFromServer(context, client);
+      if (server_time.IsSuccessful()) {
+        timestamp = server_time.GetResult();
+      } else {
+        OLP_SDK_LOG_WARNING(
+            kLogTag,
+            "Failed to get time from server, system time will be used");
       }
+    }
 
-      client::OlpClient client = CreateOlpClient(settings_, boost::none);
-      std::time_t timestamp = std::chrono::system_clock::to_time_t(
-          std::chrono::system_clock::now());
-
-      if (!settings_.use_system_time) {
-        auto server_time = GetTimeFromServer(context, client);
-        if (server_time.IsSuccessful()) {
-          timestamp = server_time.GetResult();
-        } else {
-          OLP_SDK_LOG_WARNING(
-              kLogTag,
-              "Failed to get time from server, system time will be used");
-        }
+    const auto request_body = GenerateClientBody(properties);
```

Comment:
But this will still use the same nonce for all retries, or?

Reply:
It does not use nonce, content is static

Reply:
Ahh, yes. Sorry I looked at it with only one eye open. Looks good then. Please adapt the commit message as it is to light for the amount of code in the patch.

Reply:
Done